File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        inordermap = {v: i for i, v in enumerate(inorder)}
        return build(0, 0, len(preorder)-1, preorder, inordermap)
    
def build(pid, l, r, preorder, inordermap):
    if l>r or pid>=len(preorder):
        return None
    n = TreeNode(preorder[pid])
    print("deb", preorder[pid])
    mid = inordermap[preorder[pid]]
    pid += 1
    n.left = build(pid, l, mid-1, preorder, inordermap)
    n.right = build(pid + mid - l, mid+1, r, preorder, inordermap)
    return n

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(!root)
            return false;
        return (root && subRoot && isSame(root, subRoot)) || isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }

    bool isSame(TreeNode* root, TreeNode* subRoot){
        if(root!=nullptr ^ subRoot!=nullptr)
            return false;
        if(!root && !subRoot)
            return true;
        return root->val == subRoot->val && isSame(root->left, subRoot->left) && isSame(root->right, subRoot->right);
    }
};
